File: apps/android/app/src/main/java/dev/flextranslate/foundation/CloudOptIn.kt
package dev.flextranslate.foundation

data class CloudCredential(
    val source: String,
    val expiresAtEpochMs: Long,
) {
    fun isEphemeral(nowEpochMs: Long): Boolean = source == "backend_ephemeral_token" && expiresAtEpochMs > nowEpochMs
}

data class CloudOptInState(
    val providerId: String,
    val userConsented: Boolean,
    val disclosureAccepted: Boolean,
    val credential: CloudCredential?,
    val networkState: String,
) {
    fun canStart(nowEpochMs: Long): Boolean =
        userConsented &&
            disclosureAccepted &&
            networkState == "online" &&
            credential?.isEphemeral(nowEpochMs) == true
}

class CloudOptInProvider(override val providerId: String) : CloudProvider {
    override fun isAvailable(consent: CloudConsent): Boolean =
        consent.enabled && consent.credentialSource == "backend_ephemeral_token"

    fun isAvailable(state: CloudOptInState, nowEpochMs: Long): Boolean = state.canStart(nowEpochMs)
}

object CloudProviderRegistry {
    val providers = listOf(
        CloudOptInProvider("cloud-stt-recognition-fallback"),
        CloudOptInProvider("gemini-live-assistant"),
        CloudOptInProvider("gemini-batch-audio-enrichment"),
    )
}

File: apps/ios/FlexTranslate/Sources/FlexTranslate/CloudOptIn.swift
import Foundation

struct CloudCredential: Equatable {
    let source: String
    let expiresAtEpochMs: Int64

    func isEphemeral(nowEpochMs: Int64) -> Bool {
        source == "backend_ephemeral_token" && expiresAtEpochMs > nowEpochMs
    }
}

struct CloudOptInState: Equatable {
    let providerId: String
    let userConsented: Bool
    let disclosureAccepted: Bool
    let credential: CloudCredential?
    let networkState: String

    func canStart(nowEpochMs: Int64) -> Bool {
        userConsented &&
            disclosureAccepted &&
            networkState == "online" &&
            (credential?.isEphemeral(nowEpochMs: nowEpochMs) == true)
    }
}

struct CloudOptInProvider: CloudProvider {
    let providerId: String

    func isAvailable(consent: CloudConsent) -> Bool {
        consent.enabled && consent.credentialSource == "backend_ephemeral_token"
    }

    func isAvailable(state: CloudOptInState, nowEpochMs: Int64) -> Bool {
        state.canStart(nowEpochMs: nowEpochMs)
    }
}

enum CloudProviderRegistry {
    static let providers = [
        CloudOptInProvider(providerId: "cloud-stt-recognition-fallback"),
        CloudOptInProvider(providerId: "gemini-live-assistant"),
        CloudOptInProvider(providerId: "gemini-batch-audio-enrichment")
    ]
}
